src/middleware/createRouter.ts

```diff
@@ -0,0 +1,92 @@
+import simpleStateMapping from '../lib/stateMappings/simple';
+import historyRouter from '../lib/routers/history';
+import { Index } from '../widgets/index/index';
+import { Middleware } from '.';
+import { Router, StateMapping, UiState } from '../types';
+
+const walk = (current: Index, callback: (index: Index) => void) => {
+  callback(current);
+  current
+    .getWidgets()
+    .filter(function(widget): widget is Index {
+      return widget.$$type === 'ais.index';
+    })
+    .forEach(innerIndex => {
+      walk(innerIndex, callback);
+    });
+};
+
+interface RouterProps {
+  router?: Router;
+  stateMapping?: StateMapping;
+}
+
+export type RoutingManager = (props?: RouterProps) => Middleware;
+
+export const createRouter: RoutingManager = (props = {}) => {
+  const {
+    router = historyRouter(),
+    stateMapping = simpleStateMapping(),
+  } = props;
+
+  return ({ instantSearchInstance }) => {
+    function topLevelCreateURL(nextState: UiState) {
+      const uiState = Object.keys(nextState).reduce(
+        (acc, indexId) => ({
+          ...acc,
+          [indexId]: nextState[indexId],
+        }),
+        instantSearchInstance.mainIndex.getWidgetState({})
+      );
+
+      const route = stateMapping.stateToRoute(uiState);
+
+      return router.createURL(route);
+    }
+
+    instantSearchInstance._createURL = topLevelCreateURL;
+    instantSearchInstance._initialUiState = {
```

Comment:
this makes sense, but I don't see it in the previous implementation, what am I missing?

Reply:
It used to be here:

https://github.com/algolia/instantsearch.js/blob/49ffca2d781d2125250bdcb111d2f1dc0c43e740/src/lib/InstantSearch.ts#L258-L262

It's now part of the middleware. Note that this is a side effect though, so not that great.

Reply:
Could work for now but like you said this is not great. Maybe we could introduce more handlers to address these cases later?

Reply:
Yes, I'll think it through. What do you mean by "handlers"?

We could still replicate the logic in the original `InstantSearch.ts` but I was reluctant to it at first because I didn't want this entry to be dependant on the router. The technical solution would be to mutate those in the constructor by importing the basic `stateMapping` and reading it from here. But I wanted it completely detached so that we don't need to import this package if we don't use it. What I'm aiming here is to make the package more easily modular so people load only the parts of the code they need. Do you think that's an OK tradeoff?

Reply:
Sorry if I wasn't clear, I was thinking about some additional methods on the object returned by the middleware. Something like this (not tested, just for the idea):

```js
function router({ instantSearchInstance }) {
  return {
    onStateChange() {},
    subscribe() {},
    unsubscribe() {},
    getInitialUiState() {},
  };
}
```

```js
this._initialUiState = this.middleware.reduce((initialUiState, m) => {
  return m.getInitialUiState(initialUiState);
}, {});
```

Reply:
I see. Ideally I think we shouldn't give the capability of setting the initial UI state in the middleware though.

I renamed the method `EXEPRIMENTAL_use` for now. We can reiterate on this once we roll it out and find more use cases.